File: src/offer26.SubstructureInTree/substructure_in_tree.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
        bool res = false;
        if (pRoot1 != nullptr && pRoot2 != nullptr){
            if (pRoot1->val == pRoot2->val){
                res = HasSubtreeCore(pRoot1, pRoot2);
            }
            if (!res){
                res = HasSubtree(pRoot1->left, pRoot2);
            }
            if (!res){
                res = HasSubtree(pRoot1->right, pRoot2);
            }
        }
        return res;
    }
    bool HasSubtreeCore(TreeNode* pRoot1, TreeNode* pRoot2){
        if (pRoot2 == nullptr){
            return true;
        }
        if (pRoot1 == nullptr){
            return false;
        }
        if (pRoot1->val != pRoot2->val){
            return false;
        }
        return HasSubtreeCore(pRoot1->left, pRoot2->left) && HasSubtreeCore(pRoot1->right, pRoot2->right);
    }
};

File: src/offer26.SubstructureInTree/substructure_in_tree.java
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    public boolean HasSubtree(TreeNode root1,TreeNode root2) {
        boolean res = false;
        if(root1 != null && root2 != null){
            if(root1.val == root2.val){
                res = HasSubtreeCore(root1, root2);
            }
            if(!res){
                res = HasSubtree(root1.left, root2);
            }
            if(!res){
                res = HasSubtree(root1.right, root2);
            }
        }
        return res;
    }
    private boolean HasSubtreeCore(TreeNode root1, TreeNode root2){
        if(root2 == null){
            return true;
        }
        if(root1 == null){
            return false;
        }
        if(root1.val != root2.val){
            return false;
        }
        return HasSubtreeCore(root1.left, root2.left) && HasSubtreeCore(root1.right, root2.right);
    }
}